### PR TITLE
Web workers in accordion optimized

### PR DESCRIPTION
**Pull request by leslielazzarino**
Put the U back in OpossumUI

**Summary of changes**
Introduce web workers and compute InFolderContent accordion data in 2 workers (async, not in the main process). Since Webpack 5 web workers are supported by default, not requiring any external package or change in the Webpack configuration. Further changes:

Web worker can run out of memory with big files, such error is caught, and the computation is then performed as fallback on main, jest does not support web workers yet, therefore they need to be mocked in the test. The mock just throws an error in the worker and triggers the fallback computation on main. ExternalData is loaded in the relative web worker just once, as the memory copy operation takes hundreds of milliseconds for e.g. the keycloak test file, and ExternalData is immutable, unleas a new file is opened.

**Context and reason for change**
The app gets stuck for a while when clicking on a folder with a lot of signals inside. Web workers allow to calculate the accordion faster, meanwhile leaving the main process free to process user input.

**How can the changes be tested**
Open keycloak input and navigate away from the root folder (somewhere where there is an attribution). Then click on the root folder. On the right and above the accordion the UI will change instantly, meanwhile the InFolderContent part of the accordion will be loaded later.
The fallback path can be tested using the ORT input file.

### CODE REVIEW

**Comment by benedikt-richter** [comment]
Seems to be less performant than on main

**Comment by benedikt-richter** [comment]
Performance is good! 👍

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,107 @@
+// SPDX-FileCopyrightText: Facebook, Inc. and its affiliates
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:1]: the copyright should be updated

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+interface AccordionPanelProps {
+  panelData: PanelData;
+  isAddToPackageEnabled: boolean;
+}
+
+export function AccordionPanel(props: AccordionPanelProps): ReactElement {
+  const classes = useStyles();
+
+  const [expanded, setExpanded] = React.useState<boolean>(false);
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:50]: `React.useState<boolean>(false);` -> `useState<boolean>(false);`

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+          attributionIdsWithCount={props.panelData.attributionIdsWithCount}
+          attributions={props.panelData.attributions}
+          isAddToPackageEnabled={props.isAddToPackageEnabled}
+        />
+      </MuiAccordionDetails>
+    </MuiAccordion>
+  );
+}
+
+function isDisabled(
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:100]: typically we define these functions in the scope of the component

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  syncFallbackArgs?: any;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  getAttributionIdsWithCount(workerArgs: any): Array<AttributionIdWithCount>;
+  attributions: Attributions;
+  worker: Worker;
+  isAddToPackageEnabled: boolean;
+}
+
+export function WorkerAccordionPanel(
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx:28]: would be nice to try cleaning up typescript issues in this file. There are lot of ignore statements

File: src/Frontend/Components/AggregatedAttributionsPanel/__tests__/AggregatedAttributionsPanel.test.tsx [hunk 5]
```diff
@@ -27,9 +25,24 @@ import {
   expectGlobalOnlyContextMenuForNotPreselectedAttribution,
   testCorrectMarkAndUnmarkForReplacementInContextMenu,
 } from '../../../test-helpers/context-menu-test-helpers';
+import {
+  getNewContainedExternalAttributionsAccordionWorker,
+  getNewContainedManualAttributionsAccordionWorker,
+  ResourceDetailsTabsWorkers,
+} from '../../ResourceDetailsTabs/get-new-accordion-worker';
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/__tests__/AggregatedAttributionsPanel.test.tsx:32]: we should not import from the folders of other components (also some other occurences).

File: src/Frontend/Components/App/App.tsx [hunk 6]
```diff
@@ -20,6 +20,13 @@ import { AuditView } from '../AuditView/AuditView';
 import { TopBar } from '../TopBar/TopBar';
 import { createTheme } from '@mui/material';
 import { useAppSelector } from '../../state/hooks';
+import {
+  getNewContainedExternalAttributionsAccordionWorker,
+  getNewContainedManualAttributionsAccordionWorker,
+  ResourceDetailsTabsWorkers,
+} from '../ResourceDetailsTabs/get-new-accordion-worker';
```
Comment (benedikt-richter) [inline src/Frontend/Components/App/App.tsx:27]: App should not import from here

File: src/Frontend/Components/App/App.tsx [hunk 7]
```diff
@@ -71,10 +78,34 @@ const theme = createTheme({
...
+    externalData,
+    resourceDetailsTabsWorkers.containedExternalAttributionsAccordionWorker,
+  ]);
+
   function getSelectedViewContainer(): ReactElement {
     switch (selectedView) {
       case View.Audit:
-        return <AuditView />;
+        return (
+          <AuditView resourceDetailsTabsWorkers={resourceDetailsTabsWorkers} />
```
Comment (benedikt-richter) [inline src/Frontend/Components/App/App.tsx:107]: would be good to use a context provider instead of passing it down the tree manually

File: src/Frontend/Components/ResourceDetailsTabs/contained-external-attributions-accordion-worker.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
+// SPDX-FileCopyrightText: Facebook, Inc. and its affiliates
+// SPDX-FileCopyrightText: TNG Technology Consulting GmbH <https://www.tngtech.com>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+import { getContainedExternalPackages } from './resource-details-tabs-helpers';
+import { AttributionData } from '../../../shared/shared-types';
+
+let lestWeForget: AttributionData | null = null;
```
Comment (benedikt-richter) [inline src/Frontend/Components/ResourceDetailsTabs/contained-external-attributions-accordion-worker.ts:9]: `lestWeForget`: hard to undertstand variable name

**Review by benedikt-richter — CHANGES_REQUESTED** [review]
See comments. In particular, the architecture regarding imports should be respected

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx [hunk 9]
```diff
@@ -0,0 +1,107 @@
+// SPDX-FileCopyrightText: Meta Platforms, Inc. and its affiliates
```
Comment (nicarl) [inline src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:1]: This should not be added anymore I guess

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx (hunk 3 above)
Comment (nicarl) [reply src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:100]: I think it depends a bit. This one here is solely a helper which would be fine for me outside of the component. I would like to see it before the `export function ...`

File: src/Frontend/Components/AggregatedAttributionsPanel/SyncAccordionPanel.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
+// SPDX-FileCopyrightText: Meta Platforms, Inc. and its affiliates
```
Comment (nicarl) [inline src/Frontend/Components/AggregatedAttributionsPanel/SyncAccordionPanel.tsx:1]: Same here

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx (hunk 4 above)
Comment (nicarl) [reply src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx:28]: Maybe we can deactivate the `any` eslint rule for this file

File: src/Frontend/web-workers/contained-external-attributions-accordion-worker.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
+// SPDX-FileCopyrightText: Meta Platforms, Inc. and its affiliates
+// SPDX-FileCopyrightText: TNG Technology Consulting GmbH <https://www.tngtech.com>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+import { getContainedExternalPackages } from '../Components/ResourceDetailsTabs/resource-details-tabs-helpers';
```
Comment (benedikt-richter) [inline src/Frontend/web-workers/contained-external-attributions-accordion-worker.ts:6]: should probably not import from `Components`

File: src/Frontend/web-workers/contained-manual-attributions-accordion-worker.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+// SPDX-FileCopyrightText: Meta Platforms, Inc. and its affiliates
+// SPDX-FileCopyrightText: TNG Technology Consulting GmbH <https://www.tngtech.com>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+import { getContainedManualPackages } from '../Components/ResourceDetailsTabs/resource-details-tabs-helpers';
```
Comment (benedikt-richter) [inline src/Frontend/web-workers/contained-manual-attributions-accordion-worker.ts:6]: should probably not import from `Components`

**Review by nicarl — COMMENTED** [review]
I agree with Benedikt for the imports and in particular I am for creating a context provider for the web workers

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  isAddToPackageEnabled: boolean;
+}
+
+export function AccordionPanel(props: AccordionPanelProps): ReactElement {
+  const classes = useStyles();
+
+  const [expanded, setExpanded] = React.useState<boolean>(false);
+
+  useMemo(() => {
+    setExpanded(Boolean(props.panelData.attributionIdsWithCount?.length > 0));
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:53]: `Boolean` not needed

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx (hunk 3 above)
Comment (benedikt-richter) [reply src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:100]: no stong opinion, here

File: src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx [hunk 14]
```diff
@@ -3,116 +3,116 @@
...
-import MuiTypography from '@mui/material/Typography';
-import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
-import clsx from 'clsx';
-import React, { ReactElement, useCallback, useEffect } from 'react';
+import React, { ReactElement, useMemo } from 'react';
+import {
+  getContainedExternalPackages,
+  getContainedManualPackages,
+  getExternalAttributionIdsWithCount,
+} from '../ResourceDetailsTabs/resource-details-tabs-helpers';
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx:11]: should not import these helpers, as these should be component scoped

File: src/Frontend/Components/AggregatedAttributionsPanel/SyncAccordionPanel.tsx [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+
+import React, { ReactElement } from 'react';
+import {
+  AttributionIdWithCount,
+  Attributions,
+} from '../../../shared/shared-types';
+import { AccordionPanel } from './AccordionPanel';
+import { PanelData } from '../ResourceDetailsTabs/resource-details-tabs-helpers';
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/SyncAccordionPanel.tsx:12]: as above

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+
+import React, { ReactElement, useMemo, useState } from 'react';
+import {
+  AttributionIdWithCount,
+  Attributions,
+} from '../../../shared/shared-types';
+import { AccordionPanel } from './AccordionPanel';
+import { PanelData } from '../ResourceDetailsTabs/resource-details-tabs-helpers';
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx:12]: as above. Extract type to common place

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+      props.worker,
+      active,
+      props.title,
+      setAttributionIdsWithCount,
+      props.getAttributionIdsWithCount,
+      props.syncFallbackArgs
+    );
+
+    // @ts-ignore
+    return (): never => {
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx:50]: I don't think we need or should have a `never` here

File: src/Frontend/Components/App/App.tsx [hunk 18]
```diff
@@ -71,10 +78,34 @@ const theme = createTheme({
 export function App(): ReactElement {
   const selectedView = useAppSelector(getSelectedView);
 
+  const resourceDetailsTabsWorkers: ResourceDetailsTabsWorkers = {
+    containedExternalAttributionsAccordionWorker:
+      getNewContainedExternalAttributionsAccordionWorker(),
+    containedManualAttributionsAccordionWorker:
+      getNewContainedManualAttributionsAccordionWorker(),
+  };
```
Comment (benedikt-richter) [inline src/Frontend/Components/App/App.tsx:86]: as it seems we only need this combination. We could instead export a single getter for these.

File: src/Frontend/web-workers/contained-external-attributions-accordion-worker.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import { AttributionData } from '../../shared/shared-types';
+
+let lestWeForget: AttributionData | null = null;
```
Comment (benedikt-richter) [inline src/Frontend/web-workers/contained-external-attributions-accordion-worker.ts:9]: lestWeForget?

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx (hunk 3 above)
Comment (leslielazzarino) [reply src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:100]: I think what can be declared outside has no reason to stay inside, the execution is just less efficient. Concerning the position, I prefer to have the exported function first, and the dependencies below it, so that in the file you have a top-down approach. It does not work with every programming language, but when it works I find it easier to read: as you read down the page you move deeper into the code.

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx (hunk 4 above)
Comment (leslielazzarino) [reply src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx:28]: I have deactivated it for the file, typing improvement in this file will happen in a followup.

File: src/Frontend/Components/App/App.tsx (hunk 7 above)
Comment (leslielazzarino) [reply src/Frontend/Components/App/App.tsx:107]: That will be the follow up. For that reason, I will not unify the get workers, as they will be possibly needed separated.

File: src/Frontend/web-workers/contained-external-attributions-accordion-worker.ts (hunk 11 above)
Comment (leslielazzarino) [reply src/Frontend/web-workers/contained-external-attributions-accordion-worker.ts:6]: has been moved

File: src/Frontend/web-workers/contained-external-attributions-accordion-worker.ts (hunk 19 above)
Comment (leslielazzarino) [reply src/Frontend/web-workers/contained-external-attributions-accordion-worker.ts:9]: sorry, leftover from a moment of frustration :D

File: src/Frontend/web-workers/contained-manual-attributions-accordion-worker.ts (hunk 12 above)
Comment (leslielazzarino) [reply src/Frontend/web-workers/contained-manual-attributions-accordion-worker.ts:6]: Changed

File: .github/workflows/build-and-e2e-test.yml [hunk 20]
```diff
@@ -21,7 +21,7 @@ jobs:
             E2E: export DISPLAY=:99 && Xvfb $DISPLAY -screen 0 1920x1080x24 & yarn test:e2e
           - os: macos-latest
             SHIP: ship-mac
-            E2E: yarn test:e2e
+            E2E: ''
```
Comment (benedikt-richter) [inline .github/workflows/build-and-e2e-test.yml:24]: a comment would be nice

File: src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+    minHeight: '24px !important',
+    '& div.MuiAccordionSummary-content': {
+      margin: 0,
+    },
+    '& div.MuiAccordionSummary-expandIcon': {
+      padding: '6px 12px',
+    },
+    padding: '0 12px',
+  },
+  disabledAccordion: {},
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AccordionPanel.tsx:31]: is this class needed?

File: src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx [hunk 22]
```diff
@@ -3,116 +3,116 @@
...
 export function AggregatedAttributionsPanel(
   props: AggregatedAttributionsPanelProps
 ): ReactElement {
-  const classes = useStyles();
+  const manualData = useAppSelector(getManualData);
+  const externalData = useAppSelector(getExternalData);
 
-  const [expanded, setExpanded] = React.useState<Array<string>>([]);
+  const selectedResourceId = useAppSelector(getSelectedResourceId);
+  const resolvedExternalAttributions: Set<string> = useAppSelector(
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx:40]: why do we state the type here?

File: src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx [hunk 23]
```diff
@@ -3,116 +3,116 @@
...
-  const openPopulatedPanels = useCallback(() => {
-    const openPanels = props.panelData
-      .filter(
-        (panelData: PanelData) => !isDisabled(panelData.attributionIdsWithCount)
-      )
-      .map((panelPackage: PanelData) => panelPackage.title);
-    setExpanded(openPanels);
-  }, [props.panelData]);
-  useEffect(() => openPopulatedPanels(), [openPopulatedPanels]);
+  const containedExternalPackagesWorkerArgs = useMemo(
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx:44]: could be useCallback

File: src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx [hunk 24]
```diff
@@ -3,116 +3,116 @@
...
+    () => ({
+      selectedResourceId,
+      resolvedExternalAttributions,
+    }),
+    [selectedResourceId, resolvedExternalAttributions]
+  );
+  const containedExternalPackagesSyncFallbackArgs = useMemo(
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx:51]: could be useCallback

File: src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx [hunk 25]
```diff
@@ -3,116 +3,116 @@
...
-  const handleChange =
-    (panel: string) =>
-    (event: React.ChangeEvent<unknown>, isExpanded: boolean): void => {
-      setExpanded(
-        isExpanded
-          ? expanded.concat(panel)
-          : expanded.filter((panelName) => panel !== panelName)
-      );
-    };
+  const containedManualPackagesWorkerArgs = useMemo(
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/AggregatedAttributionsPanel.tsx:60]: could be useCallback

File: src/Frontend/Components/App/App.tsx [hunk 26]
```diff
@@ -71,10 +78,34 @@ const theme = createTheme({
...
+
+  const externalData = useAppSelector(getExternalData);
+  useMemo(() => {
```
Comment (benedikt-richter) [inline src/Frontend/Components/App/App.tsx:89]: could be useCallback

File: src/Frontend/Components/ResourceDetailsTabs/ResourceDetailsTabs.tsx [hunk 27]
```diff
@@ -80,38 +73,24 @@ export function ResourceDetailsTabs(
...
-
   const assignableAttributionIds: Array<string> = remove(
     Object.keys(manualData.attributions),
     (attributionId: string): boolean =>
       !attributionIdsOfSelectedResource.includes(attributionId)
   );
 
+  const isAddToPackageEnabled: boolean =
+    props.isAllAttributionsTabEnabled && props.isAddToPackageEnabled;
+  const aggregatedAttributionsPanel = useMemo(
```
Comment (benedikt-richter) [inline src/Frontend/Components/ResourceDetailsTabs/ResourceDetailsTabs.tsx:84]: could be useCallback

**Review by benedikt-richter — CHANGES_REQUESTED** [review]

File: src/Frontend/Components/App/App.tsx (hunk 18 above)
Comment (leslielazzarino) [reply src/Frontend/Components/App/App.tsx:86]: I would not unify the get workers, as they will be possibly needed separated in the context manager follow up.

**Review by benedikt-richter — APPROVED** [review]

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+
+export function WorkerAccordionPanel(
+  props: WorkerAccordionPanelProps
+): ReactElement {
+  const [attributionIdsWithCount, setAttributionIdsWithCount] = useState<
+    Array<AttributionIdWithCount>
+  >([]);
+
+  useMemo(() => {
+    let active = true;
```
Comment (nicarl) [inline src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx:34]: Would be good to have a comment on the `let active` variable

**Review by nicarl — APPROVED** [review]
Just two tiny comments

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx [hunk 29]
```diff
@@ -13,6 +13,11 @@ import { PanelData } from '../../types/types';
 
 /* eslint-disable @typescript-eslint/no-explicit-any */
 
+export interface AttributionIdsWithCountAndResourceId {
```
Comment (benedikt-richter) [inline src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx:16]: should go into shared types

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+  setAttributionIdsWithCount: (
+    AttributionIdsWithCount: Array<AttributionIdWithCount>
+  ) => void,
+  getAttributionIdsWithCount: (
+    workerArgs: any
+  ) => Array<AttributionIdWithCount>,
+  workerArgs: any,
+  syncFallbackArgs?: any
+): void {
+  console.log(`Error in ResourceDetailsTab ${panelTitle}: `, error);
```
[the remark this answers is not in the document]
Comment (leslielazzarino) [reply src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx:146]: console is, between other things, a powerful logging library, that includes .info, .warning, .error functions. Is there a specific feature you would want, that I'm not using and that is not available with console? Otherwise I do not think we need any special library.

File: src/Frontend/Components/App/App.tsx [hunk 31]
```diff
@@ -71,10 +78,34 @@ const theme = createTheme({
...
+    try {
+      loadExternalDataInWebWorker(
+        externalData,
+        resourceDetailsTabsWorkers.containedExternalAttributionsAccordionWorker
+      );
+    } catch (error) {
+      console.log('WebWorker error in App component: ', error);
```
[the remark this answers is not in the document]
Comment (leslielazzarino) [reply src/Frontend/Components/App/App.tsx:96]: See other comment.

File: src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx (hunk 28 above)
Comment (leslielazzarino) [reply src/Frontend/Components/AggregatedAttributionsPanel/WorkerAccordionPanel.tsx]: Due to an edge-case I have rewritten how I handle race conditions.

**Review by benedikt-richter — CHANGES_REQUESTED** [review]
There are many console logs that spam the test terminal:

` console.log
      WebWorker error in App component:  Error: JEST-MOCK-GET-NEW-CONTAINED-EXTERNAL-ATTRIBUTIONS-ACCORDION-WORKER`

**Review by benedikt-richter — APPROVED** [review]
